spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/filter/factory/JsonToGrpcGatewayFilterFactory.java

```diff
@@ -0,0 +1,314 @@
+ *
+ * Making it transparent for the consumer that the service under the gateway is a gRPC
+ * one.
+ *
+ * @author Alberto C. Ríos
+ */
+public class JsonToGrpcGatewayFilterFactory
+		extends AbstractGatewayFilterFactory<JsonToGrpcGatewayFilterFactory.Config> {
+
+	private final GRPCSSLContext sslContext;
+
+	private final ResourceLoader resourceLoader;
+
+	public JsonToGrpcGatewayFilterFactory(GRPCSSLContext sslContext, ResourceLoader resourceLoader) {
+		super(Config.class);
+		this.sslContext = sslContext;
+		this.resourceLoader = resourceLoader;
+	}
+
+	@Override
+	public List<String> shortcutFieldOrder() {
+		return Arrays.asList("protoDescriptor", "protoFile", "service", "method");
+	}
+
+	@Override
+	public GatewayFilter apply(Config config) {
+		GatewayFilter filter = new GatewayFilter() {
+			@Override
+			public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+				GRPCResponseDecorator modifiedResponse = new GRPCResponseDecorator(exchange, config);
+
+				ServerWebExchangeUtils.setAlreadyRouted(exchange);
+				return modifiedResponse.writeWith(exchange.getRequest().getBody())
+						.then(chain.filter(exchange.mutate().response(modifiedResponse).build()));
+			}
+
+			@Override
+			public String toString() {
+				return filterToStringCreator(JsonToGrpcGatewayFilterFactory.this).toString();
+			}
+		};
+
+		int order = NettyWriteResponseFilter.WRITE_RESPONSE_FILTER_ORDER - 1;
+		return new OrderedGatewayFilter(filter, order);
+	}
+
+	public static class Config {
+
+		private String protoDescriptor;
+
+		private String protoFile;
+
+		private String service;
+
+		private String method;
+
+		public String getProtoDescriptor() {
+			return protoDescriptor;
+		}
+
+		public Config setProtoDescriptor(String protoDescriptor) {
+			this.protoDescriptor = protoDescriptor;
+			return this;
+		}
+
+		public String getProtoFile() {
+			return protoFile;
+		}
+
+		public Config setProtoFile(String protoFile) {
+			this.protoFile = protoFile;
+			return this;
+		}
+
+		public String getService() {
+			return service;
+		}
+
+		public Config setService(String service) {
+			this.service = service;
+			return this;
+		}
+
+		public String getMethod() {
+			return method;
+		}
+
+		public Config setMethod(String method) {
+			this.method = method;
+			return this;
+		}
+
+	}
+
+	class GRPCResponseDecorator extends ServerHttpResponseDecorator {
+
+		private final ServerWebExchange exchange;
+
+		private final Descriptors.Descriptor descriptor;
+
+		private final ObjectWriter objectWriter;
+
+		private final ObjectReader objectReader;
+
+		private final ClientCall<DynamicMessage, DynamicMessage> clientCall;
+
+		private final ObjectNode objectNode;
+
+		GRPCResponseDecorator(ServerWebExchange exchange, Config config) {
+			super(exchange.getResponse());
+			this.exchange = exchange;
+			try {
+				File descriptorFile = resourceLoader.getResource(config.getProtoDescriptor()).getFile();
```

Comment:
I'd change this to `Resource` and then use `Resource.getInputStream()` rather than `Files.newInputStream()`

Reply:
Done, @spencergibb  thanks!